src/sentry/static/sentry/app/utils/tokenizeSearch.tsx

```diff
@@ -1,93 +1,176 @@
-          tokenState = 'query';
-        } else {
-          tokenState = 'tags';
+      let trailingParen: string = '';
+      if (token.endsWith(')')) {
+        const parenMatch = token.match(/\)+$/g);
+        if (parenMatch) {
+          trailingParen = parenMatch[0];
+          token = token.replace(/\)+$/g, '');
         }
-        break;
       }
+
+      if (tokenState === TokenType.QUERY) {
+        this.addQuery(token);
+      } else if (tokenState === TokenType.TAG) {
+        this.addStringTag(token);
+      }
+
+      if (trailingParen !== '') {
+        this.addOp(trailingParen);
+      }
+    }
+  }
+
+  formatString() {
+    const formattedTokens: string[] = [];
+    for (const token of this.tokens) {
+      switch (token.type) {
+        case TokenType.TAG:
+          if (token.value === '' || token.value === null) {
+            formattedTokens.push(`${token.key}:""`);
+          } else if (/[\s\(\)\\"]/g.test(token.value)) {
+            formattedTokens.push(`${token.key}:"${escapeDoubleQuotes(token.value)}"`);
+          } else {
+            formattedTokens.push(`${token.key}:${token.value}`);
+          }
+          break;
+        default:
+          formattedTokens.push(token.value);
+      }
+    }
+    return formattedTokens.join(' ').trim();
+  }
+
+  addStringTag(value: string) {
+    const [key, tag] = formatTag(value);
+    this.addTag(key, [tag]);
+  }
+
+  addTag(key: string, tags: string[]) {
+    for (const t of tags) {
+      this.tagValues[key] = Array.isArray(this.tagValues[key])
+        ? [...this.tagValues[key], t]
+        : [t];
+      const token: Token = {type: TokenType.TAG, key, value: t};
+      this.tokens.push(token);
     }
-    searchParams[tokenState].push(token);
   }
 
-  const results: QueryResults = {
-    query: searchParams.query.map(formatQuery),
-  };
+  setTag(key: string, tags: string[]) {
+    this.removeTag(key);
+    this.addTag(key, tags);
+  }
+
+  getTags(key: string) {
+    return this.tagValues[key];
+  }
+
+  removeTag(key: string) {
+    this.tokens = this.tokens.filter(token => token.key !== key);
+    delete this.tagValues[key];
+  }
+
+  addQuery(value: string) {
+    const token: Token = {type: TokenType.QUERY, value: formatQuery(value)};
+    this.tokens.push(token);
+  }
+
+  addOp(value: string) {
+    const token: Token = {type: TokenType.OP, value};
+    this.tokens.push(token);
+  }
+
+  get query(): string[] {
+    return this.tokens.filter(t => t.type === TokenType.QUERY).map(t => t.value);
+  }
+
+  set query(values: string[]) {
+    this.tokens = this.tokens.filter(t => t.type !== TokenType.QUERY);
+    for (const v of values) {
+      this.addQuery(v);
+    }
+  }
 
-  for (const tag of searchParams.tags) {
-    const [key, value] = formatTag(tag);
-    results[key] = Array.isArray(results[key]) ? [...results[key], value] : [value];
+  copy() {
+    const q = new QueryResults([]);
+    q.tagValues = {...this.tagValues};
+    q.tokens = [...this.tokens];
```

Comment:
Do you need to deep clone the objects in this list? Javascript won't clone the token objects so if any are mutated they'll update the original object as well.

Reply:
There are a couple places where the query results were being used that were doing explicit cloning of the original object, so I added this to preserve that behaviour. No one modifies the tokens, just the token list.

Reply:
For example in membersFilter.tsx